### ISSUE TITLE
Run environment modified for build/link dependencies

### ISSUE THREAD

**Issue by michaelkuhn** [LLNL/spack#42095]
### Steps to reproduce

Since #42032 has been merged, loading qt-base+opengl overrides `CC` etc. due to its build/link dependency on llvm. (cc @haampie)

```console
$ . share/spack/setup-env.sh 
$ echo $CC

$ spack load gcc
$ echo $CC
$SPACK/opt/spack/linux-centos8-x86_64_v3/gcc-8.5.0/gcc-12.3.0-we6drltpmzemgqsy4yyw755rff6ljjjg/bin/gcc
$ spack load qt-base
$ echo $CC
$SPACK/opt/spack/linux-centos8-x86_64_v3/gcc-12.3.0/llvm-17.0.4-skkfll5vdw3jaiatmfombhuaelfgs3wd/bin/clang
```

### Information on your system

* **Spack:** 0.22.0.dev0 (d9b0c4ee806500aa7fd1238d71a4c66f4b03a140)
* **Python:** 3.6.8
* **Platform:** linux-centos8-x86_64_v3
* **Concretizer:** clingo

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have searched the issues of this repo and believe this is not a duplicate
- [X] I have run the failing commands in debug mode and reported the output

**Comment by haampie** [comment]
The same would have already happened with `module load` when autoloading deps.

I'm not sure what the best way to deal with this is.

**Comment by michaelkuhn** [comment]
Yes, it also happens with modules. It's just surprising that `CC`/`CXX`/etc. change like this and this already broke a build for us that involved Qt.

**Comment by haampie** [comment]
> broke a build

Outside of spack? Spack's build system shouldn't be affected by those runtime variables (except in the window of one month or so)

**Comment by michaelkuhn** [comment]
Yes, outside of Spack. Our students/researchers typically load packages (via Modules or Spack) and then build their own software.

**Comment by haampie** [comment]
Okay, I'm still not sure what's best then. We discussed in the meeting today that it would make sense to extend `setup_run_environment` with an arg of the effective deptype (reduced over in-edges or smth), so you can avoid setting `CC` / `CXX` when llvm is only a link dep.

But for modules this can't be done dynamically.

Alternatively use `llvm~clang`?

**Comment by michaelkuhn** [comment]
> Alternatively use `llvm~clang`?

I'm trying, but doing this via `packages.yaml` is still broken: #40753
I'll try to find a usable workaround.
